playground/CustomResources/CustomResources.AppHost/TestResource.cs

```diff
@@ -0,0 +1,73 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using Aspire.Hosting.Dashboard;
+using Aspire.Hosting.Lifecycle;
+
+static class TestResourceExtensions
+{
+    public static IResourceBuilder<TestResource> AddTestResource(this IDistributedApplicationBuilder builder, string name)
+    {
+        builder.Services.AddLifecycleHook<TestResourceLifecycleHook>();
+
+        var rb = builder.AddResource(new TestResource(name))
+                      .WithDashboardState(state => state with
+                      {
+                          ResourceType = "Test Resource",
+                          State = "Starting",
+                          Properties = [
+                              ("P1", "P2"),
+                              (DashboardKnownProperties.Source, "Custom")
+                          ]
+                      })
+                      .ExcludeFromManifest();
+
+        return rb;
+    }
+}
+
+internal sealed class TestResourceLifecycleHook : IDistributedApplicationLifecycleHook, IAsyncDisposable
+{
+    private readonly CancellationTokenSource _tokenSource = new();
```

Comment:
Needed because the BeforeStartAsync token is no longer canceled once the task completes?

Reply:
Right.